### Question 1: 

**Install Spark and PySpark** 

- Install Spark
- Run PySpark
- Create a local spark session
- Execute spark.version.

What's the output?
- 3.3.2
- 2.1.4
- 1.2.3
- 5.4
</br></br>

In [3]:
!pip install findspark


[notice] A new release of pip is available: 23.0 -> 23.0.1
[notice] To update, run: pip install --upgrade pip


In [4]:
import findspark
findspark.init('/opt/spark')
import pyspark
sc = pyspark.SparkContext()
import pandas as pd
from pyspark.sql import SparkSession
from pyspark.sql import functions as F
from pyspark.sql import types as t

23/03/07 10:19:01 WARN Utils: Your hostname, LAPTOP-028FK3OH resolves to a loopback address: 127.0.1.1; using 192.168.26.176 instead (on interface eth0)
23/03/07 10:19:01 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address


Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


23/03/07 10:19:03 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


In [5]:
spark = SparkSession.builder \
        .master("local[*]") \
        .getOrCreate()
spark

### Question 2: 

**HVFHW June 2021**

Read it with Spark using the same schema as we did in the lessons.</br> 
We will use this dataset for all the remaining questions.</br>
Repartition it to 12 partitions and save it to parquet.</br>
What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.</br>


- 2MB
- 24MB
- 100MB
- 250MB
</br></br>

In [8]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz

--2023-03-07 10:22:05--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz
Resolving github.com (github.com)... 140.82.121.4
Connecting to github.com (github.com)|140.82.121.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/4564ad9e-a6da-4923-ad6f-35ff02446a51?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230307%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230307T082205Z&X-Amz-Expires=300&X-Amz-Signature=33fa188eddffd535a5afac6b26d29de0846acb739a102e49fca932764cf58f32&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhvhv_tripdata_2021-06.csv.gz&response-content-type=application%2Foctet-stream [following]
--2023-03-07 10:22:06--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/4564ad9e

In [12]:
schema = t.StructType([
    t.StructField('dispatching_base_num', t.StringType(), True), 
    t.StructField('pickup_datetime', t.TimestampType(), True), 
    t.StructField('dropoff_datetime', t.TimestampType(), True), 
    t.StructField('PULocationID', t.IntegerType(), True), 
    t.StructField('DOLocationID', t.IntegerType(), True), 
    t.StructField('SR_Flag', t.StringType(), True), 
    t.StructField('Affiliated_base_number', t.StringType(), True)
])

df = spark.read \
        .schema(schema) \
        .option("header", "true") \
        .csv('fhvhv_tripdata_2021-06.csv.gz')

In [13]:
df.show(10)


+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B02764|2021-06-01 00:02:41|2021-06-01 00:07:46|         174|          18|      N|                B02764|
|              B02764|2021-06-01 00:16:16|2021-06-01 00:21:14|          32|         254|      N|                B02764|
|              B02764|2021-06-01 00:27:01|2021-06-01 00:42:11|         240|         127|      N|                B02764|
|              B02764|2021-06-01 00:46:08|2021-06-01 00:53:45|         127|         235|      N|                B02764|
|              B02510|2021-06-01 00:45:42|2021-06-01 01:03:33|         144|         146|      N|                  null|
|              B02510|2021-06-01 00:18:1

In [14]:
df.repartition(12) \
    .write \
    .parquet('fhvhv/2021/06/', mode='overwrite')

23/03/07 10:24:28 WARN MemoryManager: Total allocation exceeds 95.00% (1,020,054,720 bytes) of heap memory
Scaling row group sizes to 95.00% for 8 writers
23/03/07 10:24:28 WARN MemoryManager: Total allocation exceeds 95.00% (1,020,054,720 bytes) of heap memory
Scaling row group sizes to 84.44% for 9 writers
23/03/07 10:24:28 WARN MemoryManager: Total allocation exceeds 95.00% (1,020,054,720 bytes) of heap memory
Scaling row group sizes to 76.00% for 10 writers
23/03/07 10:24:28 WARN MemoryManager: Total allocation exceeds 95.00% (1,020,054,720 bytes) of heap memory
Scaling row group sizes to 69.09% for 11 writers
23/03/07 10:24:28 WARN MemoryManager: Total allocation exceeds 95.00% (1,020,054,720 bytes) of heap memory
Scaling row group sizes to 63.33% for 12 writers
23/03/07 10:24:33 WARN MemoryManager: Total allocation exceeds 95.00% (1,020,054,720 bytes) of heap memory
Scaling row group sizes to 69.09% for 11 writers
23/03/07 10:24:33 WARN MemoryManager: Total allocation exceeds 95.

In [15]:
!tree -h fhvhv/2021/06/

fhvhv/2021/06/
├── [ 24M]  part-00000-5f8a3c77-ec31-48bb-bb00-d7aac3394193-c000.snappy.parquet
├── [ 24M]  part-00001-5f8a3c77-ec31-48bb-bb00-d7aac3394193-c000.snappy.parquet
├── [ 24M]  part-00002-5f8a3c77-ec31-48bb-bb00-d7aac3394193-c000.snappy.parquet
├── [ 24M]  part-00003-5f8a3c77-ec31-48bb-bb00-d7aac3394193-c000.snappy.parquet
├── [ 24M]  part-00004-5f8a3c77-ec31-48bb-bb00-d7aac3394193-c000.snappy.parquet
├── [ 24M]  part-00005-5f8a3c77-ec31-48bb-bb00-d7aac3394193-c000.snappy.parquet
├── [ 24M]  part-00006-5f8a3c77-ec31-48bb-bb00-d7aac3394193-c000.snappy.parquet
├── [ 24M]  part-00007-5f8a3c77-ec31-48bb-bb00-d7aac3394193-c000.snappy.parquet
├── [ 24M]  part-00008-5f8a3c77-ec31-48bb-bb00-d7aac3394193-c000.snappy.parquet
├── [ 24M]  part-00009-5f8a3c77-ec31-48bb-bb00-d7aac3394193-c000.snappy.parquet
├── [ 24M]  part-00010-5f8a3c77-ec31-48bb-bb00-d7aac3394193-c000.snappy.parquet
├── [ 24M]  part-00011-5f8a3c77-ec31-48bb-bb00-d7aac3394193-c000.snappy.parquet
└── [   0]  _SUCCESS

0 d

Answer: **24MB**

### Question 3: 

**Count records**  

How many taxi trips were there on June 15?</br></br>
Consider only trips that started on June 15.</br>

- 308,164
- 12,856
- 452,470
- 50,982
</br></br>

In [17]:
df = spark.read.parquet('fhvhv/2021/06/*')
df.filter(F.to_date('pickup_datetime') == '2021-06-15').count()

452470

Answer: **452,470**

### Question 4: 

**Longest trip for each day**  

Now calculate the duration for each trip.</br>
How long was the longest trip in Hours?</br>

- 66.87 Hours
- 243.44 Hours
- 7.68 Hours
- 3.32 Hours
</br></br>

In [18]:
df.withColumn('duration', F.round((F.col('dropoff_datetime').cast("long") - F.col('pickup_datetime').cast("long"))/3600,3)) \
    .orderBy(F.col('duration').desc()).show(1)

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+--------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|duration|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+--------+
|              B02872|2021-06-25 13:55:41|2021-06-28 08:48:25|          98|         265|      N|                B02872|  66.879|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+--------+
only showing top 1 row



Answer: **66.879** Hours

### Question 5: 

**User Interface**

 Spark’s User Interface which shows application's dashboard runs on which local port?</br>

- 80
- 443
- 4040
- 8080
</br></br>

Answer: **4040**

### Question 6: 

**Most frequent pickup location zone**

Load the zone lookup data into a temp view in Spark</br>
[Zone Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv)</br>

Using the zone lookup data and the fhvhv June 2021 data, what is the name of the most frequent pickup location zone?</br>

- East Chelsea
- Astoria
- Union Sq
- Crown Heights North
</br></br>

In [19]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv

--2023-03-07 10:28:31--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv
Resolving github.com (github.com)... 140.82.121.4
Connecting to github.com (github.com)|140.82.121.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6ea97ed0e6a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230307%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230307T082831Z&X-Amz-Expires=300&X-Amz-Signature=b8904e9eacd750459dec06d90e56b0dfe96df95b197cddd2adeb84dfa19f0d17&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dtaxi_zone_lookup.csv&response-content-type=application%2Foctet-stream [following]
--2023-03-07 10:28:32--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6e

In [22]:
df_zones = spark.read \
            .option("header", "true") \
            .csv('taxi_zone_lookup.csv')
df.groupBy('PULocationID') \
    .count() \
    .orderBy(F.col('count').desc()) \
    .join(df_zones, df['PULocationID'] == df_zones['LocationID'], how='left').show(1)

+------------+------+----------+--------+-------------------+------------+
|PULocationID| count|LocationID| Borough|               Zone|service_zone|
+------------+------+----------+--------+-------------------+------------+
|          61|231279|        61|Brooklyn|Crown Heights North|   Boro Zone|
+------------+------+----------+--------+-------------------+------------+
only showing top 1 row



Answer: **Crown Heights North**